packages/memory-storage/src/resource-clients/request-queue.ts

```diff
@@ -169,11 +170,21 @@ export class RequestQueueClient extends BaseClient implements storage.RequestQue
 
         const items = [];
 
-        for (const storageEntry of existingQueueById.requests.values()) {
+        const seenRequestIds = new Set<string>();
+
+        for (const requestId of [...[...this.forefrontRequestIds].reverse(), ...existingQueueById.requests.keys()]) {
```

Comment:
Low-hanging optimization: this array will always only be `limit` long, constructing the whole thing is unreasonable. Iterators for the win?

Reply:
Would be fun to make it a iterator, and then you can also iterate from end to start without cloning+reversing

Reply:
if its small enough, lets leave that for later (or never, lol). but i also don't like to see array copies, not even one, and here we have two, just to be able to iterate things.

Reply:
> if its small enough

That, sadly, only depends on the usage. Better safe than sorry; I just added a private function for iterating over those lists.

Reply:
Slightly related to this is the `filter` call for cleaning the `forefrontRequestIds` (e.g. [here](https://github.com/apify/crawlee/pull/2681/files#diff-d2cb55420a8d6e84c8c7e076df7eb96eb74243fdfb091f3836fe7646dc7337ecR261)).

There, I feel like it is warranted - the `forefrontRequestIds` list will likely be very short most of the time - and gets shorter in every step as forefront requests are processed first.